### PR TITLE
Generate man pages from the Command description

### PR DESCRIPTION
**Pull request by dnephin**
`spf13/cobra` includes support for generating a man page. It makes use of `cpuguy83/go-md2man` which we already use to generate the existing man pages from markdown.

The cobra generated man page is very similar to the output of our current generation. There are a few small differences. If any of them are considered to be significant, we can update the generation code to output what we want.

Notable differences:
- the `SYNOPSIS` string only says `OPTIONS` instead of the full list of flags
- the command group page (`docker volume`, `docker network`, etc) lists the child commands under "SEE ALSO" on a single line, instead of listing them out one-per-line in a "COMMANDS" section
- the date is displayed in the footer of the page (where it currently says "Docker User Manuals") instead of in the header

By keeping the command description and examples in the code it should make it easier to keep the man pages up-to-date and accurate. It makes it easier to add new commands, as we can generate the man pages automatically without adding any new files. We can also use this `Long` (description) text to generate the command reference used by the online docs.

The existing markdown files have not been removed.  They will override the generated files if one exists for the same command. If the design of this PR is accepted I can start moving the text for more of the commands into the code.

This PR makes the following changes:
- update the `man/Dockerfile` to include all the new required dependencies
- update the packaging scripts to call the new `man/generate.sh` to generate all the man pages
- remove the files for `volume` commands, and move the description into the `Command.Long` field.
- adds man pages for all the new 1.12 commands. These man pages will be a little bare until we add a `Long` description to the commands.

cc @cpuguy83, @icecrime

### CODE REVIEW

**Comment by LK4D4** [comment]
@dnephin failures are real

**Comment by LK4D4** [comment]
@dnephin Do you know why we need go-md2man in main Dockerfile?

**Comment by LK4D4** [comment]
LGTM

**Comment by vdemeester** [comment]
hum, `make manpages` failed for me :

```
[ERROR] Failed to set version on github.com/spf13/cobra to dc45219961f875acff5ee07ed263e5dc19e0c5f1: exit status 128
[INFO] Setting version for github.com/spf13/viper to c1ccc378a054ea8d4e38d8c67f6938d4760b53dd.
[INFO] Setting version for golang.org/x/sys to 62bee037599929a6e9146f29d10dd5208c43507d.
[INFO] Setting version for gopkg.in/yaml.v2 to a83829b6f1293c91addabc89d0571c246397bbf4.
An Error has occurred
The command '/bin/sh -c glide install && mv vendor src' returned a non-zero code: 2
```

**Comment by dnephin** [comment]
I was waiting for the spf13/cobra PR (which included some bug fixes for this) to get merged before tagging. I has been merged now, so I'll update the sha here.

**Comment by dnephin** [comment]
Fixed the tag. I'm using `v1.3` as the version now.

**Comment by albers** [comment]
I think the representation of options with arguments in the synopsis is wrong.
Given that expressions inside of brackets are optional, the general syntax of an option with a mandatory argument should be

    [--option[=]value]

in order to describe both legal syntax variants `--option value` and `--option=value`.

Here is an example from the generated man pages:

    [--net[="bridge"]]

This means that `--net=bridge` and `--net` are both valid argument expressions, which is not true:

```bash
$ docker run --net
flag needs an argument: --net
```

The wrong syntax is also used in most current static man pages, so I understand why the generated output picked up this syntax. This is a good moment to tidy this up.

**Comment by albers** [comment]
Please ignore my previous comment.
I did not notice that only selected commands have been proted so far and picked a non-generated one.

File: api/client/volume/inspect.go [hunk 1]
```diff
@@ -44,3 +45,11 @@ func runInspect(dockerCli *client.DockerCli, opts inspectOptions) error {
 
 	return inspect.Inspect(dockerCli.Out(), opts.names, opts.format, getVolFunc)
 }
+
+var inspectDescription = `
+Returns information about one or more volumes. By default, this command renders
+all results in a JSON array. You can specify an alternate format to execute a
+given template is executed for each result. Go's http://golang.org/pkg/text/template/
```
Comment (thaJeztah) [inline api/client/volume/inspect.go]: https:// ?

Comment (dnephin) [inline api/client/volume/inspect.go]: Fixed, this was just copied from the old markdown man page

**Comment by icecrime** [comment]
That's really awesome :+1: Design LGTM (cc @tianon @SvenDowideit who might be interested too).

**Comment by SvenDowideit** [comment]
LGTM - we'll want to consume this into the cli docs eventually too

**Comment by thaJeztah** [comment]
I'm still a bit on the fence on this one. I'm happy with it being generated
automatically, but on the other hand, there's some things "attached" to it.
I was writing down some things, but not finished, so I'll paste my quick notes here;

- making "documentation" PR's becomes more difficult; to make
  changes to this part of the documentation now requires contributors
  to touch the actual code
- related to the above; we still have to maintain the online documentation
  /reference (which has a lot of the same content). Can we still copy/paste
  between the Markdown files and the embedded docs (looking at it,
  back-tics will break?)

File: api/client/volume/create.go [hunk 2]
```diff
@@ -60,3 +61,34 @@ func runCreate(dockerCli *client.DockerCli, opts createOptions) error {
 	fmt.Fprintf(dockerCli.Out(), "%s\n", vol.Name)
 	return nil
 }
+
+var createDescription = `
+Creates a new volume that containers can consume and store data in. If a name is not specified, Docker generates a random name. You create a volume and then configure the container to use it, for example:
```
Comment (thaJeztah) [inline api/client/volume/create.go]: Should we stick to 80-chars for these strings as well, to be consistent with our markdown?

Comment (LK4D4) [inline api/client/volume/create.go]: ping @dnephin

Comment (dnephin) [inline api/client/volume/create.go]: I copied this from the markdown without reformatting it. So it is consistent with the old markdown at least :)

Comment (thaJeztah) [inline api/client/volume/create.go]: We stick to 80 chars, perhaps this slipped through during review, but let's not inherit that mistake

**Comment by vdemeester** [comment]
I kinda like that but I'm wondering how to handle more complicated manpages, like [`docker-ps.1.md`](https://raw.githubusercontent.com/docker/docker/master/man/docker-ps.1.md) or [`docker-run.1.md`](https://raw.githubusercontent.com/docker/docker/master/man/docker-run.1.md) where we have longer description for some/most of the flags than what we have/describe in the code (and thus in `--help`). For example the list of possible `--filter` for `ps` (or `--format`) or the big description for `--cpu-shares` in `run`, …

There is also some manpages where we have an `EXAMPLES` section that should be taken care of.

So I'm torn :sweat_smile:. I like that but I feel we should see how to handle these more complex command manpages before doing anything.

**Comment by dnephin** [comment]
If we need additional text, we can add it to the description. It's not uncommon for man pages to have sections for things that require more explanation. If we really want a more verbose man page for some commands we can keep the `md` around for those. There is always an override from `.md` file.

The default should really be automatic generation from the source. If that's insufficient for some flags or commands we have the ability to extend the generated docs, but I think that is more the exception than the rule.

>  For example the list of possible --filter for ps

This is an excellent example. We can document that by adding the following section to the description:

```

## Filters

exited=<int>
  filter to containers which have an exit code of <int>
...
```

Putting this detail into a section in the description gives us more space to describe things.

> There is also some manpages where we have an EXAMPLES section

cobra supports the example section as well. I use it [in this PR for `volume rm`](https://github.com/docker/docker/pull/23825/files#diff-26c40b24423f4b7b8bc0679f737ec869R51).  In the end this is just another section, so it could just as easily go into the description.

> I feel we should see how to handle these more complex command manpages before doing anything.

Hopefully that covers it, but if not I'd be happy to provide more examples.

File: api/client/volume/create.go (hunk 2 above)
Comment (dnephin) [inline api/client/volume/create.go]: Cool, I fixed it

**Comment by dnephin** [comment]
@thaJeztah any outstanding concerns on this?

**Comment by thaJeztah** [comment]
@dnephin I still have concerns, but perhaps the best way is to test/try it and see if it works well in our workflow. We can evaluate after a while. So "lgtm"

**Comment by thaJeztah** [comment]
@dnephin really appreciated though, please don't take it the wrong way 👍

**Comment by icecrime** [comment]
Alright let's go for it then :innocent: LGTM

**Comment by icecrime** [comment]
Not sure if we want to cherry-pick this.

@tiborvass @thaJeztah Adding the label, but up to you!

**Pull request merged**